src/binder/clrprivbinderassemblyloadcontext.cpp

```diff
@@ -251,9 +271,51 @@ HRESULT CLRPrivBinderAssemblyLoadContext::SetupContext(DWORD      dwAppDomainId,
     return hr;
 }
 
+/* static */
+void CLRPrivBinderAssemblyLoadContext::DestroyContext(CLRPrivBinderAssemblyLoadContext *pBindContext, INT_PTR ptrManagedStrongAssemblyLoadContext)
+{
+    CONTRACTL
+    {
+        GC_NOTRIGGER;
+        THROWS;
+        MODE_PREEMPTIVE;
+        SO_TOLERANT;
+    }
+    CONTRACTL_END;
+
+    GCX_COOP();
+    // Replace the weak handle with a strong handle
+    // in order to be able to callback Unloading safely
+    OBJECTHANDLE handle = reinterpret_cast<OBJECTHANDLE>(pBindContext->m_ptrManagedAssemblyLoadContext);
+    OBJECTHANDLE strongHandle = reinterpret_cast<OBJECTHANDLE>(ptrManagedStrongAssemblyLoadContext);
+    HndDestroyHandle(HndGetHandleTable(handle), HNDTYPE_WEAK_SHORT, handle);
+    pBindContext->m_ptrManagedAssemblyLoadContext = reinterpret_cast<INT_PTR>(strongHandle);
+
+    // We cannot delete the binder here as it is used indirectly when comparing assemblies with the same binder
+    // It will be deleted by the LoaderAllocator once
+#if !defined(DACCESS_COMPILE) && !defined(CROSSGEN_COMPILE)
+    ((AssemblyLoaderAllocator*)pBindContext->m_pAssemblyLoaderAllocator)->SetBinderToRelease(pBindContext);
+#endif
+
+    pBindContext->m_pAssemblyLoaderAllocator->Release();
+    pBindContext->m_pAssemblyLoaderAllocator = NULL;
+
+    // Destroy the strong handle to the LoaderAllocator in order to let it reach its finalizer
+    DestroyHandle(reinterpret_cast<OBJECTHANDLE>(pBindContext->m_loaderAllocatorHandle));
+}
+
 CLRPrivBinderAssemblyLoadContext::CLRPrivBinderAssemblyLoadContext()
 {
     m_pTPABinder = NULL;
 }
 
+void CLRPrivBinderAssemblyLoadContext::ReleaseManagedAssemblyLoadContext()
+{
+    VERIFY(m_ptrManagedAssemblyLoadContext != NULL);
+    OBJECTHANDLE handle = reinterpret_cast<OBJECTHANDLE>(m_ptrManagedAssemblyLoadContext);
+    DestroyHandle(handle);
```

Comment:
I agree - the naming update, and comments, would be helpful here. I would suggest DestroyContext to be something like PrepareForLoadContextRelease and ReleaseManagedAssemblyLoadContext to be ReleaseLoadContext (so that, in future, additional cleanup can be added that is not related to managed instance).